Migrate templating guide (#1189)
astronomer/docs: guides/templating.md

---
title: "Use Airflow templates"

> sidebar_label: "Airflow templates"

Reviewer: ```suggestion
sidebar_label: "Jinja templates"
```

- How to validate templates.
- How to apply custom variables and functions when templating.
- How to render templates to strings and native Python code.

## Runtime variables in Airflow

Templating in Airflow works the same as Jinja templating in Python. You enclose the code you want evaluated between double curly braces, and the expression is evaluated at runtime. 

This table lists some of the most commonly used Airflow variables that you can use in templates:

| Variable name  | Description                                                |
|----------------|------------------------------------------------------------|
| execution_date | Starting datetime of DAG run interval                      |
| ds             | `execution_date` formatted as "2022-12-27"                 |
| ds_nodash      | `execution_date` formatted as "20221227"                   |
| next_ds        | next `execution_date` (= end of current interval) datetime |

For a complete list of the available variables, see the Airflow [Templates reference](https://airflow.apache.org/docs/apache-airflow/stable/macros-ref.html#default-variables).

## Templateable fields and scripts

> Templates cannot be applied to all arguments of an operator. Two attributes in the BaseOperator define limitations on templating:

Reviewer: ```suggestion
Templates cannot be applied to all arguments of an operator. Two attributes in the BaseOperator define where you can use templated values:
```

> - `template_fields`: Defines which fields are templateable.

Reviewer: ```suggestion
- `template_fields`: Defines which operator arguments can use templated values.
```

> - `template_ext`: Defines which file extensions are templateable.

Reviewer: ```suggestion
- `template_ext`: Defines which file extensions can use templated values.
```

The following example shows a simplified version of the BashOperator:

```python
class BashOperator(BaseOperator):
    template_fields = ('bash_command', 'env')  # defines which fields are templateable
    template_ext = ('.sh', '.bash')  # defines which file extensions are templateable

    def __init__(
        self,
        *,
        bash_command,
        env: None,
        output_encoding: 'utf-8',
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.bash_command = bash_command  # templateable (can also give path to .sh or .bash script)
        self.env = env  # templateable
        self.output_encoding = output_encoding  # not templateable
```

> The `template_fields` attribute holds a list of attributes that can be templated. You can also find this list in [the Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/bash/index.html#airflow.operators.bash.BashOperator.template_fields) or in the Airflow UI as shown in the following image:

Reviewer: ```suggestion
The `template_fields` attribute holds a list of attributes that can use templated values. You can also find this list in [the Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/bash/index.html#airflow.operators.bash.BashOperator.template_fields) or in the Airflow UI as shown in the following image:
```

![Rendered Template view](/img/guides/taskinstancedetails.png)

> `template_ext` contains a list of file extensions that can be read and templated at runtime. For example, instead of providing a Bash command to `bash_command`, you could provide a `.sh` script:

Reviewer: ```suggestion
`template_ext` contains a list of file extensions that can be read and templated at runtime. For example, instead of providing a Bash command to `bash_command`, you could provide a `.sh` script that contains a templated value:
```

```python
run_this = BashOperator(
    task_id="run_this",
    bash_command="script.sh",  # .sh extension can be read and templated
)
```

> The following command reads the contents of `script.sh`, templates it, and the executes it:

Reviewer: ```suggestion
The BashOperator takes the contents of the following script, templates it, and executes it:
```

```bash
# script.sh
echo "Today is {{ execution_date.format('dddd') }}"
```

Templating from files speeds development because an integrated development environment (IDE) can apply language-specific syntax highlighting on the script. This wouldn't be possible if your script is defined as a big string of Airflow code.

> By default, Airflow searches for `script.sh` relative to the directory the DAG file is defined in. So, if your DAG is stored in `/path/to/dag.py` and your script is stored in `/path/to/scripts/script.sh`, you set the value of `bash_command` to `scripts/script.sh`.

Reviewer: ```suggestion
By default, Airflow searches for the location of your scripts relative to the directory the DAG file is defined in. So, if your DAG is stored in `/path/to/dag.py` and your script is stored in `/path/to/scripts/script.sh`, you would update the value of `bash_command` in the previous example to `scripts/script.sh`.
```

> Additional search paths can be controlled at the DAG-level with the `template_searchpath` argument. For example:

Reviewer: ```suggestion
Alternatively, you can set a base path for templates at the DAG-level with the `template_searchpath` argument. For example, the following DAG would look for `script.sh` at `/tmp/script.sh`:
```

```python
with DAG(..., template_searchpath="/tmp") as dag:
    run_this = BashOperator(task_id="run_this", bash_command="script.sh")
```

## Validate templates

The output of templates can be checked in both the Airflow UI and Airflow CLI. One advantage of the Airflow CLI is that you don't need to run any tasks before seeing the result.

The Airflow CLI command `airflow tasks render` renders all templateable attributes of a given task. Given a `dag_id`, `task_id`, and dummy `execution_date`, the command output is similar to the following example:

```bash
$ airflow tasks render example_dag run_this 2021-01-01

# ----------------------------------------------------------
# property: bash_command
# ----------------------------------------------------------
echo "Today is Friday"

# ----------------------------------------------------------
# property: env
# ----------------------------------------------------------
None
```

> For this command to work, Airflow needs access to a metastore. To set up a local SQLite metastore run the following command:

Reviewer: ```suggestion
For this command to work, Airflow needs access to a metadata database. To set up a local SQLite database, run the following commands:
```

```bash

> cd [your project dir]

Reviewer: ```suggestion
cd <your-project-directory>
```

export AIRFLOW_HOME=$(pwd)
airflow db init  # generates airflow.db, airflow.cfg, and webserver_config.py in your project dir

# airflow tasks render [dag_id] [task_id] [execution_date]
```

> 

Reviewer: ```suggestion

If you use the Astro CLI, a postgres metadata database is automatically configured for you after running `astro dev start` in your project directory. From here, you can run `astro dev run tasks render <parameters>` to test your templated values. 

```

> For most templates, this is sufficient. However, if an external system such as a variable in your production Airflow metastore is reached by the templating logic, you must have connectivity to it.

Reviewer: ```suggestion
For most templates, this is sufficient. However, if an external system such as a variable in your production Airflow metadata database is reached by the templating logic, you must have connectivity to it.
```
